### DISCUSSION TITLE
Apply a style with selectors without losing your mind

### DISCUSSION THREAD

**Discussion by Kermalis** [AvaloniaUI/Avalonia#11376] [Q&A]
**Describe the bug**
I would like to apply a style to borders in a specific a window, without affecting literally all of the default avalonia controls that have borders. Even if you use the selector system, which exists purely for this purpose, it still affects borders in other controls such as NumericUpDown.

I have tried to even use "descendant" selectors which didn't work at all. The best part is that even if it did work, it would still apply to other controls since they're descendants too. So there's no way to tackle this without massive xaml bloat. And good luck trying to use this as an Application style, and then having hundreds or thousands of `Classes=""` or whatever in your program.

**To Reproduce:**
```xaml
<Window.Styles>
	<Style Selector="Border">
		<Setter Property="Margin" Value="2" />
		<Setter Property="BorderThickness" Value="1" />
		<Setter Property="BorderBrush" Value="{DynamicResource SystemAccentColor}" />
		<Setter Property="CornerRadius" Value="8,2,8,2" />
	</Style>
</Window.Styles>
```

**Expected behavior:**
The borders in my window look like that, and NumericUpDown etc are not affected.
I expect the Avalonia website's page about Selectors to even mention that this will happen, or explain how to avoid it. Does nobody else use Border selectors? I don't want to have to type a class or key a billion times in xaml. Style selectors exist to avoid that. Otherwise, why have style selectors at all

**Screenshots:**
![image](https://github.com/AvaloniaUI/Avalonia/assets/29823718/ff25b81f-3be6-430c-af48-fcc62ad571ad)

**Desktop:**
- OS: Win10
- Version: 11.0.0-preview5

**Additional context:**
It seems that this person talked about this three years ago: https://github.com/AvaloniaUI/Avalonia/issues/4801
Not sure if they're after the same behavior but it's so basic that I can't comprehend anything. Wouldn't every Avalonia user experience this or am I missing something extremely obvious, that ChatGPT and Google can't help either? I mean, I must be the dumbest person ever or something with how basic this is.

**Comment by maxkatz6** [comment]
You are overcomplicating your problem.
What you did - you created a style that applies on every border (inside of the window). Sets Margin, BorderBrush and some other properties.
What program did - exactly what you asked - applied these properties on every Border in the window.
And not surprisingly, tons of default controls are using Border inside of them as well.

What you actually need - to apply your style on some of your Borders.
To do so, use class or name selector:
```xaml
<Window.Styles>
	<Style Selector="Border.accent">
		<Setter Property="Margin" Value="2" />
		<Setter Property="BorderThickness" Value="1" />
		<Setter Property="BorderBrush" Value="{DynamicResource SystemAccentColor}" />
		<Setter Property="CornerRadius" Value="8,2,8,2" />
	</Style>
</Window.Styles>

<Border Classes="accent">
    <!-- Content -->
</Border>
```
Or
```xaml
<Window.Styles>
	<Style Selector="Border#Specific">
		<Setter Property="Margin" Value="2" />
		<Setter Property="BorderThickness" Value="1" />
		<Setter Property="BorderBrush" Value="{DynamicResource SystemAccentColor}" />
		<Setter Property="CornerRadius" Value="8,2,8,2" />
	</Style>
</Window.Styles>

<Border Name="Specific">
    <!-- Content -->
</Border>
```

The difference is how specific style is. One with class selector can be reused on multiple borders. One with name selector can be used only on one specific (in the scope).

> It seems that this person talked about this three years ago: https://github.com/AvaloniaUI/Avalonia/issues/4801

It was me, and that issue is unrelated to your problem. It was about resources defined inside of specific styles being available outside of these styles - which also by design, but not perfect.

> that ChatGPT and Google can't help either

ChatGPT doesn't know what he is saying, as it's an AI. Helpfulness of it depends on luck.
Google though would help with our documentation - https://docs.avaloniaui.net/docs/styling (and see sub pages).

**Comment by Kermalis** [comment]
I know about classes, but is there really no way to do this? It seems like such a miss to not have. Why would I want to type the class hundreds of times?

**Reply by maxkatz6** [reply]
> but is there really no way to do this

Do what? To not have what? You can create a subclass of Border (i.e. inherit from Border), apply style on it and use instead of simple Border.

But using Classes is a normal solution as well. Developers usually do more specific selectors for classes.
It's not a bloat.

**Reply by maxkatz6** [reply]
As alternatives, you could use ControlTheme and set it on each Border (like it would be done in WPF). Or define these properties on each border directly. Neither is better than classes in terms of symbols typed.

**Comment by Kermalis** [comment]
A subclass of Border is actually a good idea, probably what I would prefer to do. Thanks for the suggestion
